posts/Gunnar/2019-01-23-testing-cdi-beans-and-persistence-layer-under-java-se.adoc

```diff
@@ -0,0 +1,456 @@
+    public boolean isTransactionActive() {
+        try {
+            return com.arjuna.ats.jta.UserTransaction.userTransaction().getStatus() == Status.STATUS_ACTIVE;
+        }
+        catch (SystemException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public UserTransaction getUserTransaction() {
+        return com.arjuna.ats.jta.UserTransaction.userTransaction();
+    }
+}
+----
+
+This lets Weld
+
+* register JTA synchronizations (which is used to make transactional observer methods work),
+* query for the current transaction status and
+* obtain the user transaction (so to enable injection of the `UserTransaction` object).
+
+The `TransactionServices` implementation is picked up using the service loader mechanism,
+so a file _META-INF/services/org.jboss.weld.bootstrap.api.Service_ with the fully-qualified name of our implementation as its contents is required:
+
+[source,java]
+----
+org.hibernate.demos.jpacditesting.support.TestingTransactionServices
+----
+
+And with that, we can now test code like this which makes use of transactional observers:
+
+[source,java]
+----
+@ApplicationScoped
+public class SomeObserver {
+
+    public void observes(@Observes(during=TransactionPhase.AFTER_COMPLETION) String event) {
+        // handle event ...
+    }
+}
+----
+
+We also can use JTA's `@Transactional` annotation to benefit from declarative transaction control:
+
+[source,java]
+----
+@ApplicationScoped
+public class TransactionalGreetingService {
+
+    @Transactional(TxType.REQUIRED)
+    public String greet(String name) {
+        // ...
+    }
+}
+----
+
+When this `greet()` method is invoked, it must be run in a transaction context,
+which either has been started before or will be started if needed.
+Now, if you have used transactional CDI beans before, you might wonder where the associated method interceptor is.
+As it turns out, Narayana comes with CDI support and provides us with everything needed: method interceptors for the different transactional behaviours (`REQUIRED`, `MANDATORY` etc.)
+as well as a portable extension which registers the interceptors with the CDI container.
+
+== Configuring the Weld Initiator
+
+There's one last detail we have ignored so far, and that is how Weld will detect all the beans we need for our test,
+be it actual components under test such as `GreetingService` or testing infrastructure such as the `EntityManagerProducer`.
+The simplest is to let Weld scan the classpath itself and pick up all beans it finds.
+This is enabled by passing a new `Weld` instance to the `WeldInitiator` rule:
+
+[source,java]
+----
+public class CdiJpaTest {
+
+    @Rule
+    public WeldInitiator weld = WeldInitiator.from(new Weld())
+        .activate(RequestScoped.class)
+        .inject(this)
+        .build();
+
+    @Test
+    public void someTest() {
+        // ...
+    }
+}
+----
+
+That's very convenient, but it might cause some slowness for larger classpaths.
```

Comment:
While this is true, the main issue is mostly that it will also pick up some alternatives or interceptors on your CP that you don't really want to interfere with your test.